ecosystem/typescript/sdk/src/transaction_builder/builder_utils.ts

```diff
@@ -72,84 +72,108 @@ export function ensureBigInt(val: number | bigint | string): bigint {
 }
 
 export function serializeArg(argVal: any, argType: TypeTag, serializer: Serializer) {
+  serializeArgInner(argVal, argType, serializer, 0);
+}
+
+function serializeArgInner(argVal: any, argType: TypeTag, serializer: Serializer, depth: number) {
+  // TODO: Should check how this matches to BCS's limit on nesting
+  if (depth >= 8) {
+    throw new Error("Arguments are too nested, must be no greater than 8");
+  }
   if (argType instanceof TypeTagBool) {
     serializer.serializeBool(ensureBoolean(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU8) {
+  } else if (argType instanceof TypeTagU8) {
     serializer.serializeU8(ensureNumber(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU16) {
+  } else if (argType instanceof TypeTagU16) {
     serializer.serializeU16(ensureNumber(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU32) {
+  } else if (argType instanceof TypeTagU32) {
     serializer.serializeU32(ensureNumber(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU64) {
+  } else if (argType instanceof TypeTagU64) {
     serializer.serializeU64(ensureBigInt(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU128) {
+  } else if (argType instanceof TypeTagU128) {
     serializer.serializeU128(ensureBigInt(argVal));
-    return;
-  }
-  if (argType instanceof TypeTagU256) {
+  } else if (argType instanceof TypeTagU256) {
     serializer.serializeU256(ensureBigInt(argVal));
-    return;
+  } else if (argType instanceof TypeTagAddress) {
+    serializeAddress(argVal, serializer);
+  } else if (argType instanceof TypeTagVector) {
+    serializeVector(argVal, argType, serializer, depth);
+  } else if (argType instanceof TypeTagStruct) {
+    serializeStruct(argVal, argType, serializer, depth);
+  } else {
+    throw new Error("Unsupported arg type.");
   }
-  if (argType instanceof TypeTagAddress) {
-    let addr: AccountAddress;
-    if (typeof argVal === "string" || argVal instanceof HexString) {
-      addr = AccountAddress.fromHex(argVal);
-    } else if (argVal instanceof AccountAddress) {
-      addr = argVal;
-    } else {
-      throw new Error("Invalid account address.");
+}
+
+function serializeAddress(argVal: any, serializer: Serializer) {
+  let addr: AccountAddress;
+  if (typeof argVal === "string" || argVal instanceof HexString) {
+    addr = AccountAddress.fromHex(argVal);
+  } else if (argVal instanceof AccountAddress) {
+    addr = argVal;
+  } else {
+    throw new Error("Invalid account address.");
+  }
+  addr.serialize(serializer);
+}
```

Comment:
why not just make this part of the serializer class? for core primitives, that what I did in Python and I think Serde does something similar.

Reply:
I thought about that, and I just left it how it was for now